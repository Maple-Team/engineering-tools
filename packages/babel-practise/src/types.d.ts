declare module 'react-native-adaptive-stylesheet' {
  export function create(parameter: Record<string, any>)
}
