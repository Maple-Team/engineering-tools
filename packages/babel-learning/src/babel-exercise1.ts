import { parse } from "@babel/parser";
import generate from "@babel/generator";
import traverse from "@babel/traverse";
import { isIdentifier, isMemberExpression, numericLiteral } from "@babel/types";

const sourceCode = `
    console.log(1);

    function func() {
        console.info(2);
    }

    export default class Clazz {
        say() {
            console.debug(3);
        }
        render() {
            return <div>{ console.error(4) }</div>
        }
    }
`;
const ast = parse(sourceCode, {
  plugins: ["jsx"],
  sourceType: "unambiguous",
});
// requrement 1

if (ast) {
  traverse(ast, {
    CallExpression: (path, state) => {
      const callee = path.node.callee;
      const loc = path.node.loc;
      /**solution 1 */
      if (
        isMemberExpression(callee) &&
        isIdentifier(callee.object) &&
        callee.object.name === "console" &&
        isIdentifier(callee.property) &&
        ["log", "error", "debug", "info"].includes(callee.property.name)
      ) {
        path.node.arguments.push(
          numericLiteral(loc?.start.line || 0),
          numericLiteral(loc?.start.column || 0)
        );
      }
      /**solution 2 */
      // if (targetCalleeName.includes(generate(callee).code)) {
      //   path.node.arguments.push(numericLiteral(loc?.start.line || 0), numericLiteral(loc?.start.column || 0))
      // }
      /**solution 3 */
      // console.log(path.get('callee').toString()) console.log console.info console.debug console.error
    },
  });
  const { code } = generate(ast);
}
