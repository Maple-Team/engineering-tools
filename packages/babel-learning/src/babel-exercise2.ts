import { parse } from "@babel/parser";
import generate from "@babel/generator";
import traverse from "@babel/traverse";
import template from "@babel/template";
import { arrayExpression } from "@babel/types";

const sourceCode = `
    console.log(1);

    function func() {
        console.info(2);
    }

    export default class Clazz {
        say() {
            console.debug(3);
        }
        render() {
            return <div>{ console.error(4) }</div>
        }
    }
`;
const targetCalleeName = ["log", "info", "error", "debug"].map(
  (item) => `console.${item}`
);

// requrement 2 同一行打印会影响原本的参数的展示, 因此改为在`console.xx`节点之前打印
const ast = parse(sourceCode, {
  plugins: ["jsx"],
  sourceType: "unambiguous",
});
traverse(ast, {
  CallExpression: (path) => {
    // @ts-ignore
    if (path.node.isNew) {
      //
      return;
    }
    if (targetCalleeName.includes(generate(path.node.callee).code)) {
      const loc = path.node.loc;
      const newNode = template.expression(
        `console.log("filename: (${loc?.start.line}, ${loc?.start.column})")`
      )();
      //@ts-ignore
      newNode.isNew = true; // 跳过新节点处理的标记
      if (path.findParent((p) => p.isJSXElement())) {
        path.replaceWith(arrayExpression([newNode, path.node]));
        path.skip(); // 跳过子节点处理
      } else {
        path.insertBefore(newNode);
      }
    }
  },
});
const { code } = generate(ast);
