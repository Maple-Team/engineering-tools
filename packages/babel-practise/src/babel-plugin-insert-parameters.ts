import generate from '@babel/generator'
import template from '@babel/template'
import { arrayExpression } from '@babel/types'

const targetCalleeName = ['log', 'info', 'error', 'debug'].map((item) => `console.${item}`)

export default (): babel.PluginObj => {
  return {
    name: '@babel/plugin-insert-parameters',
    visitor: {
      CallExpression: (path, state) => {
        // @ts-ignore
        if (path.node.isNew) {
          return
        }
        if (targetCalleeName.includes(generate(path.node.callee).code)) {
          const loc = path.node.loc
          const stringArr = state.filename?.split('/') || []
          const filename = stringArr[stringArr?.length - 1]
          const directory = stringArr[stringArr?.length - 2]
          const newNode = template.expression(
            `console.log("${directory}/${filename}: Ln ${loc?.start.line}, Col ${loc?.start.column})")`
          )()
          //@ts-ignore
          newNode.isNew = true // 跳过新节点处理的标记
          if (path.findParent((p) => p.isJSXElement())) {
            path.replaceWith(arrayExpression([newNode, path.node]))
            path.skip() // 跳过子节点处理
          } else {
            path.insertBefore(newNode)
          }
        }
      },
    },
  }
}
