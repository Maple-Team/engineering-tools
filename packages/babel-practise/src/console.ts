import { parse } from '@babel/parser'
import generate from '@babel/generator'
import traverse from '@babel/traverse'
import template from '@babel/template'
import {
  Identifier,
  isArrayExpression,
  isExpression,
  isIdentifier,
  isMemberExpression,
  stringLiteral,
  numericLiteral,
  isExpressionStatement,
  callExpression,
  arrayExpression,
} from '@babel/types'

const sourceCode = `
    console.log(1);

    function func() {
        console.info(2);
    }

    export default class Clazz {
        say() {
            console.debug(3);
        }
        render() {
            return <div>{ console.error(4) }</div>
        }
    }
`
const targetCalleeName = ['log', 'info', 'error', 'debug'].map((item) => `console.${item}`)
{
  const ast = parse(sourceCode, {
    plugins: ['jsx'],
    sourceType: 'unambiguous',
  })
  // requrement 1

  if (ast) {
    traverse(ast, {
      CallExpression: (path, state) => {
        const callee = path.node.callee
        const loc = path.node.loc
        /**solution 1 */
        if (
          isMemberExpression(callee) &&
          isIdentifier(callee.object) &&
          callee.object.name === 'console' &&
          isIdentifier(callee.property) &&
          ['log', 'error', 'debug', 'info'].includes(callee.property.name)
        ) {
          path.node.arguments.push(numericLiteral(loc?.start.line || 0), numericLiteral(loc?.start.column || 0))
        }
        /**solution 2 */
        // if (targetCalleeName.includes(generate(callee).code)) {
        //   path.node.arguments.push(numericLiteral(loc?.start.line || 0), numericLiteral(loc?.start.column || 0))
        // }
        /**solution 3 */
        // console.log(path.get('callee').toString()) console.log console.info console.debug console.error
      },
    })
    const { code } = generate(ast)
    console.log(code)
  }
}

{
  // requrement 2 同一行打印会影响原本的参数的展示, 因此改为在`console.xx`节点之前打印
  const ast = parse(sourceCode, {
    plugins: ['jsx'],
    sourceType: 'unambiguous',
  })
  traverse(ast, {
    CallExpression: (path) => {
      // @ts-ignore
      if (path.node.isNew) {
        //
        return
      }
      if (targetCalleeName.includes(generate(path.node.callee).code)) {
        const loc = path.node.loc
        const newNode = template.expression(`console.log("filename: (${loc?.start.line}, ${loc?.start.column})")`)()
        //@ts-ignore
        newNode.isNew = true // 跳过新节点处理的标记
        if (path.findParent((p) => p.isJSXElement())) {
          path.replaceWith(arrayExpression([newNode, path.node]))
          path.skip() // 跳过子节点处理
        } else {
          path.insertBefore(newNode)
        }
      }
    },
  })
  const { code } = generate(ast)
  console.log(code)
}
