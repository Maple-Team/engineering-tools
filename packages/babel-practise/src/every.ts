function every<T>(
  this: ArrayLike<T>,
  callbackfn: (item: T, index?: number, args?: T[]) => boolean,
  thisArg: ArrayLike<T>
): boolean {
  let t: ArrayLike<T> | null = null,
    k = 0
  if (this === null) {
    throw new TypeError('this is null or not defined')
  }
  let O: T[] = Object(this)
  let len = O.length >>> 0
  if (typeof callbackfn !== 'function' && Object.prototype.toString.call(callbackfn) !== '[object Function]') {
    throw new TypeError('callbackfn is not a function')
  }
  if (arguments.length > 1) {
    t = thisArg
  }
  while (k < len) {
    let kValue = null
    if (k in O) {
      kValue = O[k]
      let result = t ? callbackfn.call(t, kValue, k, O) : callbackfn(kValue, k, O)
      if (!result) {
        return false
      }
    }
    k++
  }
  return true
}
Array.prototype.every = Array.prototype.every || every
