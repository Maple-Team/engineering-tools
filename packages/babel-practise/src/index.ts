import { parse, TransformOptions } from '@babel/core'
import generate from '@babel/generator'
import traverse from '@babel/traverse'
import * as t from '@babel/types'

const options: TransformOptions = {
  ast: true,
  filename: 'babel.config.json',
  rootMode: 'root',
  presets: ['@babel/preset-env'],
}
/**
 * es6 downgrade to es5
 * @param code
 * @returns
 */
const es6to5 = (code: string | undefined | null) => {
  if (!code) return
  const ast = parse(code, options)
  if (!ast) return
  traverse(ast, {
    Program: (path) => {
      const { node } = path
      for (const directive of node.directives) {
        if (directive.value.value === 'use strict') return
      }
      path.unshiftContainer('directives', t.directive(t.directiveLiteral('use strict')))
    },
    VariableDeclaration: (path) => {
      const { node } = path
      if (node.kind === 'const' || node.kind === 'let') {
        path.node.kind = 'var'
      }
    },
    ArrowFunctionExpression: (path) => {
      let { params, body, generator, async } = path.node
      if (!t.isBlockStatement(body)) {
        const node = t.returnStatement(body)
        body = t.blockStatement([node])
      }

      path.replaceWith(t.functionExpression(t.identifier('fn'), params, body, generator, async))
    },
  })

  return generate(ast, {}, code).code
}
/**
 * es5 upgrade to es6
 * @param code
 */
const es5to6 = (code: string): string => {
  return ''
}
/**
 * 操作符替换
 * @param input
 * @param target
 * @param replace
 */
const replaceOperator = (input: string, target: string, replace: string): string => {
  const ast = parse(input, options)
  if (!ast) return ''
  traverse(ast, {
    BinaryExpression: (path) => {
      let { operator, left, right } = path.node
      if (operator === target) {
        // @ts-ignore
        path.replaceWith(t.binaryExpression(replace, left, right))
      }
      //@ts-ignore
      // path.node.operator = replace
    },
  })
  return generate(ast, {}, input).code
}
const autoInjected = (code: string | undefined | null): string => {
  if (!code) return ''
  const ast = parse(code, options)
  if (!ast) return ''
  traverse(ast, {
    //
    Program: {},
    // handle import
    ImportDeclaration: {},
    ImportDefaultSpecifier: {},
    ImportNamespaceSpecifier: {},
    // inject func
    FunctionDeclaration: {},
    ClassMethod: {},
    ArrowFunctionExpression: {},
    FunctionExpression: {},
  })
  return generate(ast, {}, code).code
}
export { es6to5, replaceOperator, autoInjected }
