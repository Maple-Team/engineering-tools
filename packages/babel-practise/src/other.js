const { list, targets } = require("core-js-compat")({
  targets: "ie >= 10",
});

console.log(list, targets);

require("core-js-builder")({
  targets: ">0.5%",
  filename: "./0.5%.bundle.js",
})
  .then((code) => {
    console.log(code);
  })
  .catch(console.error);
