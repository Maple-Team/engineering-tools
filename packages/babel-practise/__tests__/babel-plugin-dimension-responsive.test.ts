import { BabelPluginDimensionResponsive } from '../src'
import type { PluginOptions } from '../src/plugins/babel-plugin-dimension-responsive'
import { transformFileSync } from '@babel/core'
import path from 'path'

describe('@babel/plugin-insert-parameter', () => {
  it('should file transformed right', () => {
    const options: PluginOptions = {
      filename: 'style.ts',
      minValue: 10,
      excludes: ['paddingTop'],
      transformfn: (value) => value + 10,
    }
    const result = transformFileSync(path.resolve(__dirname, './input/style.ts'), {
      plugins: [[BabelPluginDimensionResponsive, options]],
      parserOpts: {
        sourceType: 'unambiguous',
      },
    })
    expect(result?.code).toMatchSnapshot()
  })
})
