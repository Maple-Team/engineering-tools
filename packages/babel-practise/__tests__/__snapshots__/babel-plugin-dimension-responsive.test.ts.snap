// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@babel/plugin-insert-parameter should file transformed right 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;

var _reactNativeAdaptiveStylesheet = _interopRequireDefault(require(\\"react-native-adaptive-stylesheet\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = _reactNativeAdaptiveStylesheet.default.create({
  container: {
    backgroundColor: '#CCECDB',
    paddingHorizontal: 20
  },
  formContainer: {
    paddingHorizontal: 25,
    paddingTop: 20,
    paddingBottom: 39,
    marginBottom: 30,
    backgroundColor: '#F8F8F8',
    borderRadius: 15
  },
  loading: {
    flex: 1,
    justifyContent: 'center',
    flexDirection: 'row',
    padding: 50
  }
});

exports.default = _default;"
`;
